zbench.zig

```diff
@@ -3,421 +3,183 @@
-    /// Memory allocator used by the benchmark.
+pub const Benchmark = struct {
     allocator: std.mem.Allocator,
-    /// Configuration settings
-    config: Config,
+    common_config: Config,
+    benchmarks: std.ArrayListUnmanaged(Definition) = .{},
 
-    /// Initializes a new Benchmark instance.
-    /// name: A string representing the benchmark's name.
-    /// allocator: Memory allocator to be used.
-    pub fn init(name: []const u8, allocator: std.mem.Allocator, config: Config) !Benchmark {
-        const bench = Benchmark{
-            .name = name,
+    pub fn init(allocator: std.mem.Allocator, config: Config) Benchmark {
+        return Benchmark{
             .allocator = allocator,
-            .config = config,
-            .timer = std.time.Timer.start() catch return error.TimerUnsupported,
-            .durations = std.ArrayList(u64).init(allocator),
+            .common_config = config,
         };
-
-        return bench;
-    }
-
-    /// Starts or restarts the benchmark timer.
-    pub fn start(self: *Benchmark) void {
-        self.timer.reset();
-    }
-
-    /// Stop the benchmark and record the duration
-    pub fn stop(self: *Benchmark) void {
-        const elapsedDuration = self.timer.read();
-        self.total_duration += elapsedDuration;
-
-        if (elapsedDuration < self.min_duration) self.min_duration = elapsedDuration;
-        if (elapsedDuration > self.max_duration) self.max_duration = elapsedDuration;
-
-        self.durations.append(elapsedDuration) catch unreachable;
     }
 
-    /// Reset the benchmark
-    pub fn reset(self: *Benchmark) void {
-        self.total_operations = 0;
-        self.min_duration = std.math.maxInt(u64);
-        self.max_duration = 0;
-        self.total_duration = 0;
-        self.durations.clearRetainingCapacity();
+    pub fn deinit(self: *Benchmark) void {
+        self.benchmarks.deinit(self.allocator);
     }
 
-    /// Returns the elapsed time since the benchmark started.
-    pub fn elapsed(self: *Benchmark) u64 {
-        var sum: u64 = 0;
-        for (self.durations.items) |duration| {
-            sum += duration;
+    pub fn add(
+        self: *Benchmark,
+        name: []const u8,
+        func: *const BenchFunc,
+    ) !void {
+        try self.benchmarks.append(self.allocator, Definition{
+            .name = name,
+            .func = func,
+            .config = self.common_config,
+        });
+    }
+
+    pub fn run(self: Benchmark) !Results {
+        var arena = std.heap.ArenaAllocator.init(self.allocator);
+        defer arena.deinit();
+
+        const n_benchmarks = self.benchmarks.items.len;
+        const results = try self.allocator.alloc(Result, n_benchmarks);
+        for (self.benchmarks.items, results) |b, *r| {
+            var timings_ns: std.ArrayListUnmanaged(u64) = .{};
+            var t = try std.time.Timer.start();
+            for (0..b.config.iterations) |_| {
+                b.func(arena.allocator());
```

Comment:
Passing an arena here is probably a dumb idea, it just helped in getting some test code running quickly.